File: customer-service/src/main/java/com/wks/servicemarketplace/customerservice/config/TokenValidatorFactory.java
package com.wks.servicemarketplace.customerservice.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.wks.servicemarketplace.common.auth.StandardTokenValidator;
import com.wks.servicemarketplace.common.auth.TokenValidator;
import org.glassfish.hk2.api.Factory;

import javax.inject.Inject;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.PublicKey;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import java.util.stream.Collectors;

public class TokenValidatorFactory implements Factory<TokenValidator> {

    private final StandardTokenValidator tokenValidator;

    @Inject
    public TokenValidatorFactory(ObjectMapper objectMapper) throws Exception {
        tokenValidator = new StandardTokenValidator(getPublicKey(), objectMapper);
    }

    private PublicKey getPublicKey() {
        try (InputStreamReader inputStreamReader = new InputStreamReader(getClass().getClassLoader().getResourceAsStream("publicKey.pem"));
             BufferedReader reader = new BufferedReader(inputStreamReader)) {
            final String content = reader.lines().collect(Collectors.joining(""))
                    .replace("-----BEGIN PUBLIC KEY-----", "")
                    .replace("-----END PUBLIC KEY-----", "");
            final KeyFactory keyFactory = KeyFactory.getInstance("RSA");
            final X509EncodedKeySpec keySpecPKCS8 = new X509EncodedKeySpec(Base64.getMimeDecoder().decode(content));
            return keyFactory.generatePublic(keySpecPKCS8);
        } catch (IOException | NoSuchAlgorithmException | InvalidKeySpecException e) {
            e.printStackTrace();
            throw new RuntimeException("Failed to load public key", e);
        }
    }

    @Override
    public TokenValidator provide() {
        return tokenValidator;
    }

    @Override
    public void dispose(TokenValidator instance) {

    }
}

File: auth-service/src/main/kotlin/com/wks/servicemarketplace/authservice/config/TokenValidatorFactory.kt
package com.wks.servicemarketplace.authservice.config

import com.fasterxml.jackson.databind.ObjectMapper
import com.wks.servicemarketplace.common.auth.StandardTokenValidator
import com.wks.servicemarketplace.common.auth.TokenValidator
import org.glassfish.hk2.api.Factory
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.security.KeyFactory
import java.security.NoSuchAlgorithmException
import java.security.PublicKey
import java.security.spec.InvalidKeySpecException
import java.security.spec.X509EncodedKeySpec
import java.util.*
import java.util.stream.Collectors
import javax.inject.Inject

class TokenValidatorFactory @Inject constructor(objectMapper : ObjectMapper) : Factory<TokenValidator> {
    lateinit var tokenValidator: StandardTokenValidator
    lateinit var publicKey: PublicKey

    init {
        try {
            InputStreamReader(javaClass.classLoader.getResourceAsStream("publicKey.pem")).use { inputStreamReader ->
                BufferedReader(inputStreamReader).use { reader ->
                    val content = reader.lines().collect(Collectors.joining(""))
                            .replace("-----BEGIN PUBLIC KEY-----", "")
                            .replace("-----END PUBLIC KEY-----", "")
                    val keyFactory = KeyFactory.getInstance("RSA")
                    val keySpec = X509EncodedKeySpec(Base64.getMimeDecoder().decode(content))
                    publicKey = keyFactory.generatePublic(keySpec)
                    tokenValidator = StandardTokenValidator(publicKey, objectMapper)
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
            throw RuntimeException("Failed to load public key", e)
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
            throw RuntimeException("Failed to load public key", e)
        } catch (e: InvalidKeySpecException) {
            e.printStackTrace()
            throw RuntimeException("Failed to load public key", e)
        }
    }

    override fun provide(): TokenValidator {
        return tokenValidator
    }

    override fun dispose(instance: TokenValidator) {}
}